src/Http/Http.Results/tools/ResultsOfTGenerator/Program.cs

```diff
@@ -0,0 +1,875 @@
+        writer.WriteLine("/// </summary>");
+
+        // Class remarks doc
+        writer.WriteLine("/// <remarks>");
+        writer.WriteLine("/// An instance of this type cannot be created explicitly. Use the implicit cast operators to create an instance");
+        writer.WriteLine("/// from an instance of one of the declared type arguments, e.g.");
+        writer.WriteLine("/// <code>Results&lt;OkObjectHttpResult, ProblemHttpResult&gt; result = Results.Ok();</code>");
+        writer.WriteLine("/// </remarks>");
+
+        // Type params docs
+        for (int j = 1; j <= i; j++)
+        {
+            writer.WriteLine(@$"/// <typeparam name=""{typeArgName}{j}"">The {j.ToOrdinalWords()} result type.</typeparam>");
+        }
+
+        // Class declaration
+        writer.Write($"public sealed class {className}<");
+
+        // Type args
+        for (int j = 1; j <= i; j++)
+        {
+            writer.Write($"{typeArgName}{j}");
+            if (j != i)
+            {
+                writer.Write(", ");
+            }
+        }
+
+        writer.WriteLine("> : IResult");
+
+        // Type arg contraints
+        for (int j = 1; j <= i; j++)
+        {
+            writer.WriteIndent($"where {typeArgName}{j} : IResult");
+            if (j != i)
+            {
+                writer.WriteLine();
+            }
+        }
+        writer.WriteLine();
+        writer.WriteLine("{");
+
+        // Ctor
+        writer.WriteIndentedLine("// Use implicit cast operators to create an instance");
+        writer.WriteIndentedLine($"private {className}(IResult activeResult)");
+        writer.WriteIndentedLine("{");
+        writer.WriteIndentedLine(2, "Result = activeResult;");
+        writer.WriteIndentedLine("}");
+        writer.WriteLine();
+
+        // Result property
+        writer.WriteIndentedLine("/// <summary>");
+        writer.WriteIndentedLine($"/// Gets the actual <see cref=\"IResult\"/> returned by the <see cref=\"Endpoint\"/> route handler delegate.");
+        writer.WriteIndentedLine("/// </summary>");
+        writer.WriteIndentedLine("public IResult Result { get; }");
+        writer.WriteLine();
+
+        // ExecuteAsync method
+        writer.WriteIndentedLine("/// <inheritdoc/>");
+        writer.WriteIndentedLine("public async Task ExecuteAsync(HttpContext httpContext)");
+        writer.WriteIndentedLine("{");
+        writer.WriteIndentedLine(2, "ArgumentNullException.ThrowIfNull(httpContext, nameof(httpContext));");
+        writer.WriteLine();
+        writer.WriteIndentedLine(2, "if (Result is null)");
+        writer.WriteIndentedLine(2, "{");
+        writer.WriteIndentedLine(3, "throw new InvalidOperationException(\"The IResult assigned to the Result property must not be null.\");");
+        writer.WriteIndentedLine(2, "}");
+        writer.WriteLine();
+        writer.WriteIndentedLine(2, "await Result.ExecuteAsync(httpContext);");
```

Comment:
```suggestion
        writer.WriteIndentedLine("public Task ExecuteAsync(HttpContext httpContext)");
        writer.WriteIndentedLine("{");
        writer.WriteIndentedLine(2, "ArgumentNullException.ThrowIfNull(httpContext, nameof(httpContext));");
        writer.WriteLine();
        writer.WriteIndentedLine(2, "if (Result is null)");
        writer.WriteIndentedLine(2, "{");
        writer.WriteIndentedLine(3, "throw new InvalidOperationException(\"The IResult assigned to the Result property must not be null.\");");
        writer.WriteIndentedLine(2, "}");
        writer.WriteLine();
        writer.WriteIndentedLine(2, "return Result.ExecuteAsync(httpContext);");
```

It's a little bit more efficient if we skip the tail await at the small cost of losing the stack frame if there are any exceptions.